### PR TITLE
Python onnxruntime.seed(<seed>) API for random seed

### PR DESCRIPTION
**Pull request by jessebenson**
**Description**:
1. Add a Python API `onnxruntime.seed(<seed>)` for specifying the random seed.
_Note:  I didn't update every kernel to use this, just Dropout._  

2. Improve consistency/repeatability of Python training unit tests.

**Motivation and Context**
Add a Python API for specifying the random seed to onnxruntime, to match other frameworks, e.g.
`numpy.random.seed(<seed>)`
`torch.manual_seed(<seed>) / torch.cuda.manual_seed(<seed>)`

The `ort_trainer.py` was generating random samples directly on CUDA.  CUDA does not guarantee the same random sequence gets generated on different devices (even with the same seed).  Instead, we will generate the random sequence on CPU and move it to the device.

### CODE REVIEW

File: onnxruntime/core/framework/random_generator.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+  std::pair<uint64_t, uint64_t> GetPhiloxSeeds(uint64_t count) {
+    uint64_t offset = offset_.fetch_add(count);
+    return std::pair<uint64_t, uint64_t>(seed_, offset);
+  }
+
+  static PhiloxGenerator& Default();
+
+ private:
+  uint64_t seed_;
```
Comment (edgchen1) [inline onnxruntime/core/framework/random_generator.h:30]: why is seed_ not atomic? perhaps it would be good to document how thread-safe this class is, i.e., what methods can be called concurrently?

Comment (edgchen1) [reply onnxruntime/core/framework/random_generator.h:30]: also what's a philox? :)

File: onnxruntime/core/framework/random_seed.h [hunk 2]
```diff
@@ -12,10 +12,9 @@ namespace utils {
  * Gets the static random seed value. All calls to GetStaticRandomSeed()
```
Comment (edgchen1) [inline onnxruntime/core/framework/random_seed.h:12]: this documentation needs updating. also would suggest renaming to Get[Global?]RandomSeed()

File: onnxruntime/core/framework/random_seed.h [hunk 3]
```diff
@@ -12,10 +12,9 @@ namespace utils {
...
  * throughout the lifetime of the process will return the same value.
  *
- * @param default_seed If the random seed is not set, return this value instead of random if it's positive.
  * @return The static random seed value.
  */
-uint32_t GetStaticRandomSeed(uint32_t default_seed = 0);
+int64_t GetStaticRandomSeed();
 
 /**
  * Sets the static random seed value.
```
Comment (edgchen1) [inline onnxruntime/core/framework/random_seed.h:20]: this documentation also needs updating. suggest renaming to SetRandomSeed()

File: orttraining/orttraining/models/bert/main.cc [hunk 4]
```diff
@@ -367,6 +361,15 @@ Status ParseArguments(int argc, char* argv[], BertParameters& params, OrtParamet
     ORT_RETURN_IF_NOT(params.data_parallel_size > 0, "data_parallel_size must > 0");
     ORT_RETURN_IF_NOT(params.horizontal_parallel_size > 0, "horizontal_parallel_size must > 0");
 
+    int seed = flags["seed"].as<int>();
+    if (params.horizontal_parallel_size > 1 && seed <= 0) {
+      seed = 8211; // Megatron needs a random seed.
+    }
+    if (seed > 0) {
+      utils::SetStaticRandomSeed(static_cast<uint32_t>(seed));
```
Comment (edgchen1) [inline orttraining/orttraining/models/bert/main.cc:369]: static_cast not needed anymore

File: orttraining/orttraining/models/gpt2/main.cc [hunk 5]
```diff
@@ -227,6 +221,15 @@ Status ParseArguments(int argc, char* argv[], GPT2Parameters& params, OrtParamet
     ORT_RETURN_IF_NOT(params.data_parallel_size > 0, "data_parallel_size must > 0");
     ORT_RETURN_IF_NOT(params.horizontal_parallel_size > 0, "horizontal_parallel_size must > 0");
 
+    int seed = flags["seed"].as<int>();
+    if (params.horizontal_parallel_size > 1 && seed <= 0) {
+      seed = 8211; // Megatron needs a random seed.
+    }
+    if (seed > 0) {
+      utils::SetStaticRandomSeed(static_cast<uint32_t>(seed));
```
Comment (edgchen1) [inline orttraining/orttraining/models/gpt2/main.cc:229]: static_cast not needed anymore

File: onnxruntime/core/framework/random_generator.h (hunk 1 above)
Comment (jessebenson) [reply onnxruntime/core/framework/random_generator.h:30]: Good questions.  I'll add documentation.

[Philox](https://docs.scipy.org/doc/numpy/reference/random/bit_generators/philox.html) is a particular random number generator technique.  CUDA (curand specifically) uses it, which is how we implement the CUDA Dropout kernel.

Comment (edgchen1) [reply onnxruntime/core/framework/random_generator.h:77]: thanks for adding documentation. regarding the thread-safety, i wonder if both seed_ and offset_ need to be updated together atomically? consider concurrent calls to SetSeed() and NextPhiloxSeeds():

SetSeed() | NextPhiloxSeeds()
-|-
seed_ = seed; ||
|| uint64_t offset = offset_.fetch_add(count);
offset_.store(0); ||
|| return std::pair<uint64_t, uint64_t>(seed_, offset);

could we get an inconsistent pair?

Comment (jessebenson) [reply onnxruntime/core/framework/random_generator.h:77]: Yes we can.  Generally speaking, seed is set at start of process and then fixed after that.  But you're right this isn't strictly guaranteed.  I was trying to avoid using mutex but, I'll use a mutex just for safety.

**Comment by pengwa** [comment]
LGTM, are you planning for a convergence test? since you mentioned "CUDA does not guarantee the same random sequence gets generated on different devices (even with the same seed). Instead, we will generate the random sequence on CPU and move it to the device.", we will have better convergence curve with the change, am I right?

**Comment by jessebenson** [comment]
@skottmckay @pranavsharma - for training, we're adding a Python API to set the random seed.  Currently, I only connected it to Dropout.  I'm looking to connect it to the `random.cc` operators (RandomNormal, etc).

Each random operator currently has a `std::default_random_engine` (initialized with seed attribute or clock) plus a mutex.  After I change it, each random operator would use the new `RandomGenerator` class to generate a seed, then create a `std::default_random_engine` in each Compute call.  There's modest cost to create an engine (need to generate 624 values), but are these random operators performance sensitive?

**Comment by jessebenson** [comment]
@pengwa - good question.  The "CPU random sequence generation" here is in ORTTrainer, which it uses to generate samples for converting the model and unit tests only.  So I don't believe this will affect convergence.

File: onnxruntime/core/framework/random_generator.h (hunk 1 above)
Comment (edgchen1) [reply onnxruntime/core/framework/random_generator.h:82]: might be possible to do some fancier stuff like packing two values into a single atomic one, but i suppose until there's a demonstrated need, a mutex is just fine.
nit: now maybe we don't need to #include <atomic> anymore

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 6]
```diff
@@ -351,6 +352,9 @@ void addGlobalMethods(py::module& m, const Environment& env) {
   m.def(
       "get_device", []() -> std::string { return BACKEND_DEVICE; },
       "Return the device used to compute the prediction (CPU, MKL, ...)");
+  m.def(
```
Comment (thiagocrepaldi) [inline onnxruntime/python/onnxruntime_pybind_state.cc:355]: How does this new global seed correlate/interact with the seed recently added into ORTTrainer frontend on PR https://github.com/microsoft/onnxruntime/pull/3209?

Do we still need a seed on ORTTrainer frontend now that a global seed is being added?

Comment (jessebenson) [reply onnxruntime/python/onnxruntime_pybind_state.cc:355]: ORTTrainer seed is removed in favor of this API.

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 7]
```diff
@@ -351,6 +352,9 @@ void addGlobalMethods(py::module& m, const Environment& env) {
...
+      "seed", [](const int64_t seed) { utils::SetRandomSeed(static_cast<uint64_t>(seed)); },
```
Comment (thiagocrepaldi) [inline onnxruntime/python/onnxruntime_pybind_state.cc:356]: static_cast<uint64_t> not needed, right? Lambda binding is already enforcing it to be int64_t, as the SetRandomSeed signature requires it

File: orttraining/orttraining/core/graph/graph_augmenter.cc [hunk 8]
```diff
@@ -17,8 +17,14 @@ Status GraphAugmenter::AugmentGraph(Graph& graph, const GraphDefs& graph_element
     }
   }
 
-  // Add new nodes to the graph.
-  for (const auto& node_def : graph_element_defs.NodeDefs()) {
+  // Add new nodes to the graph, sorted by name to get a consistent ordering.
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/core/graph/graph_augmenter.cc:20]: In which case is the ordering important? Is it related to seeding or this is an opportunistic fix?

File: orttraining/orttraining/models/bert/main.cc [hunk 9]
```diff
@@ -367,6 +361,15 @@ Status ParseArguments(int argc, char* argv[], BertParameters& params, OrtParamet
     ORT_RETURN_IF_NOT(params.data_parallel_size > 0, "data_parallel_size must > 0");
     ORT_RETURN_IF_NOT(params.horizontal_parallel_size > 0, "horizontal_parallel_size must > 0");
 
+    int seed = flags["seed"].as<int>();
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/models/bert/main.cc:364]: should it be int64_t?

File: orttraining/orttraining/models/gpt2/main.cc [hunk 10]
```diff
@@ -227,6 +221,15 @@ Status ParseArguments(int argc, char* argv[], GPT2Parameters& params, OrtParamet
     ORT_RETURN_IF_NOT(params.data_parallel_size > 0, "data_parallel_size must > 0");
     ORT_RETURN_IF_NOT(params.horizontal_parallel_size > 0, "horizontal_parallel_size must > 0");
 
+    int seed = flags["seed"].as<int>();
+    if (params.horizontal_parallel_size > 1 && seed <= 0) {
+      seed = 8211; // Megatron needs a random seed.
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/models/gpt2/main.cc:226]: what about utils::GetRandomSeed() instead of 8211?

File: orttraining/orttraining/models/bert/main.cc [hunk 11]
```diff
@@ -367,6 +361,15 @@ Status ParseArguments(int argc, char* argv[], BertParameters& params, OrtParamet
...
+    if (params.horizontal_parallel_size > 1 && seed <= 0) {
+      seed = 8211; // Megatron needs a random seed.
+    }
+    if (seed > 0) {
+      utils::SetRandomSeed(seed);
+      printf("Random seed is set to %d.\n", seed);
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/models/bert/main.cc:370]: NIT: Should we use std::cout instead of printf? I noticed that tehre are existing printf, but I am not sure whether there is a guideline where we should upgrade new logs to std:;cout instead

File: orttraining/orttraining/models/bert/main.cc (hunk 9 above)
Comment (jessebenson) [reply orttraining/orttraining/models/bert/main.cc]: Seems reasonable.  I've switched these to int64_t.

File: orttraining/orttraining/models/gpt2/main.cc (hunk 10 above)
Comment (jessebenson) [reply orttraining/orttraining/models/gpt2/main.cc:226]: `utils::GetRandomSeed()` by default uses the clock for the seed (until `utils::SetRandomSeed()` is called).  So when running distributed jobs (i.e. multiple processes), each process will have a different seed.

When running with horizontal parallelism (aka Megatron-style model splitting), we are running with multiple processes.  However, we need to ensure all processes use the same seed for Dropout.

So the intent of the code here is to ensure we get a globally agreed upon initial seed, either from the user or some static value (arbitrarily 8211).

File: orttraining/orttraining/core/graph/graph_augmenter.cc (hunk 8 above)
Comment (jessebenson) [reply orttraining/orttraining/core/graph/graph_augmenter.cc:20]: More of an opportunistic fix, to ensure all distributed processes execute same sequence.  I actually don't think it's causing issues here, since from any given tensor's perspective it's computed with the same order of operations.  I think I'm going to split this change out, and measure its effects separately.

**Comment by thiagocrepaldi** [comment]
As you mentioned before, pytorch exposes 2 APIs for seed, one for CPU and another for CUDA: torch.manual_seed(<seed>) and torch.cuda.manual_seed(<seed>)

Is there a reason why we wouldnt do that for ORT too?

It is true that we normally would use ORT for high performance GPU training, only, but consider a CI test on a machine without GPU. Would we still be able to test `utils::set_seed` or the lack of CUDA would cause set_seed to fail?

**Comment by jessebenson** [comment]
I think it's better to have a single `onnxruntime.set_seed()` API rather than one per execution provider.  This API won't fail on CPU machines.  These "random generators" are not really execution provider specific - they are simple classes.
